cardano-cli/src/Cardano/CLI/Read.hs

```diff
@@ -764,7 +774,77 @@ readVoteFile
   -> VoteFile In
   -> IO (Either VoteError (VotingEntry era))
 readVoteFile w fp =
-  first VoteErrorFile <$> conwayEraOnwardsConstraints w (readFileTextEnvelope AsVotingEntry fp)
+  conwayEraOnwardsConstraints w
+    $ first VoteErrorFile <$> readFileTextEnvelope AsVotingEntry fp
+
+emptyVotingProcedures :: VotingProcedures era
+emptyVotingProcedures = VotingProcedures $ Ledger.VotingProcedures Map.empty
+
+fromVoterRole :: ()
+  => Ledger.EraCrypto (ShelleyLedgerEra era) ~ StandardCrypto
+  => ShelleyBasedEra era
+  -> Ledger.Voter (Ledger.EraCrypto (ShelleyLedgerEra era))
+  -> Voter era
+fromVoterRole _ = \case
+  Ledger.CommitteeVoter cred ->
+    VoterCommittee (VotingCredential (Ledger.coerceKeyRole cred))    -- TODO: Conway era - We shouldn't be using coerceKeyRole.
+  Ledger.DRepVoter cred ->
+    VoterDRep (VotingCredential cred)
+  Ledger.StakePoolVoter kh ->
+    VoterSpo (StakePoolKeyHash kh)
+
+-- TODO Conway delete this where we aren't using TxVotes anymore
+votingProceduresToTxVotes :: forall era. ConwayEraOnwards era -> VotingProcedures era -> TxVotes era
+votingProceduresToTxVotes w apiVps =
+  conwayEraOnwardsConstraints w $
+    case Map.toList (Ledger.unVotingProcedures (unVotingProcedures apiVps)) >>= reKey . first (fromVoterRole (conwayEraOnwardsToShelleyBasedEra w)) of
+      [] -> TxVotesNone
+      vps -> TxVotes w $ Map.fromList $ bimap (second GovernanceActionId) VotingProcedure <$> vps
+  where
+    reKey :: (a, Map.Map k v) -> [((a, k), v)]
+    reKey (a, m) = map (\(k, v) -> ((a, k), v)) $ Map.toList m
+
+singletonVotingProcedures :: ()
+  => ShelleyBasedEra era
+  -> Ledger.Voter (Ledger.EraCrypto (ShelleyLedgerEra era))
+  -> Ledger.GovActionId (Ledger.EraCrypto (ShelleyLedgerEra era))
+  -> Ledger.VotingProcedure (ShelleyLedgerEra era)
+  -> VotingProcedures era
+singletonVotingProcedures _ voter govActionId votingProcedure =
+  VotingProcedures
+    $ Ledger.VotingProcedures
+    $ Map.singleton voter
+    $ Map.singleton govActionId votingProcedure
+
+mergeVotingProcedures :: ()
+  => VotingProcedures era
+  -> VotingProcedures era
+  -> VotingProcedures era
+mergeVotingProcedures vpsa vpsb =
+  VotingProcedures
+    $ Ledger.VotingProcedures
+    $ Map.unionWith (Map.unionWith const)
+        (Ledger.unVotingProcedures (unVotingProcedures vpsa))
+        (Ledger.unVotingProcedures (unVotingProcedures vpsb))
```

Comment:
The above three functions can be moved into `cardano-api`.  Having it here proves it works and is useful.

Reply:
We can leave them here in this PR but we should move them to `cardano-api` so others can use them without depending on `cardano-api`.